alpa/pipeline_parallel/stage_construction.py

```diff
@@ -82,7 +82,130 @@ def get_last_dp_result():
             last_autosharding_option_dicts)
 
 
-@maybe_numba_jit
+@maybe_numba_jit(fastmath=True, nogil=True, cache=True)
+def get_optimal_submeshes(best_s, f_argmin, num_devices, num_layers, submesh_n_devices):
+    current_s = best_s
+    current_layer = 0
+    current_devices = num_devices
+
+    res = []
+    while current_s > 0 and current_layer < num_layers and current_devices > 0:
+        next_start_layer, submesh_choice, autosharding_choice = (
+            f_argmin[current_s, current_layer, current_devices])
+        assert next_start_layer != -1 and current_devices != -1
+        res.append(((current_layer, next_start_layer), submesh_choice,
+                    autosharding_choice))
+        current_s -= 1
+        current_layer = next_start_layer
+        current_devices -= submesh_n_devices[submesh_choice]
+    assert current_s == 0 and current_layer == num_layers and current_devices == 0
+
+    return res
+
+
+@maybe_numba_jit(fastmath=True, nogil=True, cache=True)
+def dp_impl_2(
+    num_layers, num_devices, submesh_sizes, valid_idxs_and_costs, max_n_succ_stages
+):
+    f = np.full(
+        (num_layers + 1, num_layers + 1, num_devices + 1), np.inf, dtype=np.float32
+    )
+    f_stage_max = np.full(
+        (num_layers + 1, num_layers + 1, num_devices + 1), 0.0, dtype=np.float32
+    )
+    f_argmin = np.full(
+        (num_layers + 1, num_layers + 1, num_devices + 1, 3), -1, dtype=np.int32
+    )
+    f[0, num_layers, 0] = 0
+    for d in maybe_prange(1, num_devices + 1):
```

Comment:
There are dependencies between the iterations on different values of `d`. Can we actually use `prange` here? Even if the result is correct, can `prange` here actually accelerate the execution?

Reply:
you know you're right - there's a clear loop carried dependency (duh) so this shouldn't give numerically correct answer but it does?

>Even if the result is correct, can prange here actually accelerate the execution?

if the question is whether `prange` is actually faster, the answer is yes i think - at the minimum it is actually running a bunch of processes in parallel (you can check `htop`). whether there's some kind of synchronization cost that negates the benefit of the parallelization i don't know (i didn't carefully profile the difference with/without). either way it's not mathematically correct so i'll take it out, but it might be worth thinking about whether the inherent structure of the optimization somehow breaks the dependence (and that's why it's not coming up incorrect).